SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestMultipleOwnership/SpatialTestMultipleOwnership.cpp

```diff
@@ -0,0 +1,213 @@
+			{
+				return;
+			}
+
+			// Preemptively empty the Pawns array so that we don't get any issues with index-based addressing later on in the test
+			MultipleOwnershipPawns.Empty();
+
+			for (AActor* Pawn : FoundPawns)
+			{
+				AMultipleOwnershipPawn* MultipleOwnershipPawn = Cast<AMultipleOwnershipPawn>(Pawn);
+				if (MultipleOwnershipPawn != nullptr)
+				{
+					MultipleOwnershipPawns.Add(MultipleOwnershipPawn);
+				}
+			}
+			MultipleOwnershipPawns.Sort([](AMultipleOwnershipPawn& LHS, AMultipleOwnershipPawn& RHS) -> bool {
+				return LHS.GetActorLocation().Y < RHS.GetActorLocation().Y;
+			});
+			// Double checking that we have the correct number of MultipleOwnershipPawns, just to make sure that the casting above was
+			// successful and no issues arise from that
+			if (MultipleOwnershipPawns.Num() == 2)
+			{
+				FinishStep();
+			}
+		},
+		5.0f);
+
+	// Step definition for Client 1 to send a Server RPC which is used multiple times during the test. RPCs are expected to miss twice on
+	// the first pass, then once, then 0 times..
+	FSpatialFunctionalTestStepDefinition ClientSendRPCStepDefinition;
+	ClientSendRPCStepDefinition.StepName = TEXT("ClientSendRPCsTest");
+	ClientSendRPCStepDefinition.bIsNativeDefinition = true;
+	ClientSendRPCStepDefinition.NativeStartEvent.BindLambda([this]() {
+		for (AMultipleOwnershipPawn* MultipleOwnershipPawn : MultipleOwnershipPawns)
+		{
+			MultipleOwnershipPawn->ServerSendRPC();
+		}
+
+		FinishStep();
+	});
+
+	// Client 1 sends an RPC from both MultipleOwnershipPawns
+	AddStepFromDefinition(ClientSendRPCStepDefinition, FWorkerDefinition::Client(1));
+
+	// All workers check that both MultipleOwnershipPawns have received 0 RPCs, therefore that the RPCs were ignored since the
+	// MultipleOwnershipPawns had no owner.
+	AddStep(
+		TEXT("SpatialTestMultipleOwnershipAllWorkersCheckCorrectRPCs"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+		[this](float DeltaTime) {
+			if (MultipleOwnershipPawns[0]->ReceivedRPCs == 0 && MultipleOwnershipPawns[1]->ReceivedRPCs == 0)
+			{
+				FinishStep();
+			}
+		},
+		10.0f);
+
+	// The server makes Client 1's PlayerController possess the first MultipleOwnershipPawn.
+	AddStep(TEXT("SpatialTestMultipleOwnershipPossessPawn1"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		APlayerController* PlayerController =
+			Cast<APlayerController>(GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1)->GetOwner());
+		PlayerController->Possess(MultipleOwnershipPawns[0]);
+
+		FinishStep();
+	});
+
+	// Client 1 sends an RPC from both MultipleOwnershipPawns
+	AddStepFromDefinition(ClientSendRPCStepDefinition, FWorkerDefinition::Client(1));
+
+	auto PossessesCorrectPawns2 = [this]() {
+		APlayerController* PlayerController =
+			Cast<APlayerController>(GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1)->GetOwner());
+		return PlayerController == nullptr || PlayerController->GetPawn() == MultipleOwnershipPawns[0];
+	};
```

Comment:
Interesting, took me a while to understand what this was for.
But as it's not reused, I think we could move it into the step itself.
I guess the advantage of keeping it here is that you can name the lambda, but I think moving it to the AddStep also tells you what the lambda is for, as it will be in the `IsReady` part, so you know it's trying to determine whether the step can proceed.
But I think I can see either viewpoint, so can probably be convinced.

Reply:
Yeah, I was thinking about making it a lambda in there, but I thought it would be neater to have it outside,  just for future readability. I could rename them to something like "RPCs2 IsReady" to make it clear?